core/drivers/crypto/se050/core/rsa.c

```diff
@@ -0,0 +1,705 @@
+					     &key_bin, NULL,
+					     crypto_bignum_num_bytes(key->n)
+					     * 8);
+	free(key_bin.e);
+	free(key_bin.d);
+	free(key_bin.n);
+	free(key_bin.p);
+	free(key_bin.q);
+	free(key_bin.qp);
+	free(key_bin.dp);
+	free(key_bin.dq);
+
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, k_object);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result decrypt_es(uint32_t algo, struct rsa_keypair *key,
+			     const uint8_t *src, size_t src_len,
+			     uint8_t *dst, size_t *dst_len)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_asymmetric_t ctx = { 0 };
+	sss_se05x_object_t kobject = { 0 };
+	TEE_Result res = TEE_SUCCESS;
+
+	res = se050_inject_keypair(&kobject, key);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	st = sss_se05x_asymmetric_context_init(&ctx, se050_session, &kobject,
+					       tee2se050(algo),
+					       kMode_SSS_Decrypt);
+	if (st != kStatus_SSS_Success) {
+		if (!se050_rsa_keypair_from_nvm(key))
+			sss_se05x_key_store_erase_key(se050_kstore, &kobject);
+
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	st = sss_se05x_asymmetric_decrypt(&ctx, src, src_len, dst, dst_len);
+	if (st != kStatus_SSS_Success)
+		res = TEE_ERROR_BAD_PARAMETERS;
+
+	if (!se050_rsa_keypair_from_nvm(key))
+		sss_se05x_key_store_erase_key(se050_kstore, &kobject);
+
+	sss_se05x_asymmetric_context_free(&ctx);
+
+	return res;
+}
+
+static TEE_Result decrypt_nopad(struct rsa_keypair *key, const uint8_t *src,
+				size_t src_len, uint8_t *dst, size_t *dst_len)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_asymmetric_t ctx = { 0 };
+	sss_se05x_object_t kobject = { 0 };
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *buf = NULL;
+	size_t offset = 0;
+	size_t blen = 0;
+	size_t rsa_len = 0;
+
+	res = se050_inject_keypair(&kobject, key);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	st = sss_se05x_asymmetric_context_init(&ctx, se050_session, &kobject,
+					       kAlgorithm_SSS_RSASSA_NO_PADDING,
+					       kMode_SSS_Decrypt);
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, &kobject);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	blen = CFG_CORE_BIGNUM_MAX_BITS / 8;
+	buf = malloc(blen);
+	if (!buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	rsa_len = crypto_bignum_num_bytes(key->n);
+	memset(buf, 0, blen);
+	memcpy(buf + rsa_len - src_len, src, src_len);
+
+	st = sss_se05x_asymmetric_decrypt(&ctx, buf, src_len, buf, &blen);
+	if (st != kStatus_SSS_Success)
+		res = TEE_ERROR_BAD_PARAMETERS;
+
+	/* Remove the zero-padding (leave one zero if buff is all zeroes) */
+	offset = 0;
+	while ((offset < rsa_len - 1) && (buf[offset] == 0))
+		offset++;
+
+	if (*dst_len < rsa_len - offset) {
+		*dst_len = rsa_len - offset;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto out;
+	}
+	*dst_len = rsa_len - offset;
+	memcpy(dst, (char *)buf + offset, *dst_len);
+out:
+	free(buf);
```

Comment:
call `free(buf)` if buf not NULL

Reply:
other maintainers recommend just the opposite. could you double check please?

Reply:
ok just calling free of buf NULL is doing unnecessary operation like (lock/unlock spinlock), up to you

Reply:
Not checking `buf` is perfectly fine and is the recommended style. The rationale is simplicity, we normally don't care about saving a function call, and if the extra lock/unlock is found to be a problem it can be avoided in `bget_malloc.c`.

Reply:
Fine :)